kpatch-build/create-diff-object.c

```diff
@@ -321,6 +321,93 @@ static int kpatch_warn_only_change(struct section *sec)
 	return 1;
 }
 
+/*
+ * Determine if a section has changed only due to might_sleep's
+ * embedding of the line number into an instruction operand.
+ *
+ * Example:
+ *
+ *   50f:   be f7 01 00 00          mov    $0x1f7,%esi
+ *   514:   48 c7 c7 00 00 00 00    mov    $0x0,%rdi
+ *                          517: R_X86_64_32S       .rodata.do_select.str1.8+0x98
+ *   51b:   e8 00 00 00 00          callq  520 <do_select+0x520>
+ *                          51c: R_X86_64_PC32      ___might_sleep-0x4
+ */
+static int kpatch_debug_atomic_sleep_change(struct section *sec)
+{
+	struct insn insn1, insn2;
+	unsigned long start1, start2, size, offset, length;
+	struct rela *rela;
+	int warnonly = 0, found;
+
+	if (sec->status != CHANGED ||
+	    is_rela_section(sec) ||
+	    !is_text_section(sec) ||
+	    sec->sh.sh_size != sec->twin->sh.sh_size ||
+	    !sec->rela ||
+	    sec->rela->status != SAME)
+		return 0;
+
+	start1 = (unsigned long)sec->twin->data->d_buf;
+	start2 = (unsigned long)sec->data->d_buf;
+	size = sec->sh.sh_size;
+	for (offset = 0; offset < size; offset += length) {
+		insn_init(&insn1, (void *)(start1 + offset), 1);
+		insn_init(&insn2, (void *)(start2 + offset), 1);
+		insn_get_length(&insn1);
+		insn_get_length(&insn2);
+		length = insn1.length;
+
+		if (!insn1.length || !insn2.length)
+			ERROR("can't decode instruction in section %s at offset 0x%lx",
+			      sec->name, offset);
+
+		if (insn1.length != insn2.length)
+			return 0;
+
+		if (!memcmp((void *)start1 + offset, (void *)start2 + offset,
+			    length))
+			continue;
+
+		/* verify it's a mov immediate to %esi */
+		insn_get_opcode(&insn1);
+		insn_get_opcode(&insn2);
+		if (insn1.opcode.value != 0xbe || insn2.opcode.value != 0xbe)
+			return 0;
+
+		/* verify the next instruction is mov $0x0,%rdi */
+		if (*(unsigned int *)(start1 + offset + length) != 0xc7c748)
+			return 0;
```

Comment:
Thanks for double checking, you're right about the code pattern.
**block/blk-core.c :: generic_make_request_checks()** is an interesting example of this:
```
0000000000000000 <generic_make_request_checks>:
   0:   0f 1f 44 00 00          nopl   0x0(%rax,%rax,1)
                        1: R_X86_64_NONE        __fentry__-0x4
   5:   55                      push   %rbp
   6:   31 d2                   xor    %edx,%edx
>  8:   be 60 07 00 00          mov    $0x760,%esi
   d:   48 89 e5                mov    %rsp,%rbp
  10:   41 57                   push   %r15
  12:   41 56                   push   %r14
  14:   41 55                   push   %r13
  16:   41 54                   push   %r12
  18:   49 89 ff                mov    %rdi,%r15
  1b:   53                      push   %rbx
  1c:   48 83 ec 38             sub    $0x38,%rsp
  20:   8b 5f 28                mov    0x28(%rdi),%ebx
> 23:   48 c7 c7 00 00 00 00    mov    $0x0,%rdi
                        26: R_X86_64_32S        .rodata.blk_start_queue.str1.1
  2a:   65 48 8b 04 25 28 00    mov    %gs:0x28,%rax
  31:   00 00
  33:   48 89 45 d0             mov    %rax,-0x30(%rbp)
  37:   31 c0                   xor    %eax,%eax
> 39:   e8 00 00 00 00          callq  3e <generic_make_request_checks+0x3e>
                        3a: R_X86_64_PC32       __might_sleep-0x4
```
where the line number is stored before saving the stack frame, then the filename relocation is saved, more instructions, then finally the callq + __might_sleep rela.

If I understand **kpatch_warn_only_change()** correctly, it's verifying for a given section:
- before and after versions must match up until...
- an immediate mov to %esi.
- a "warn_slowpath_" non-string relocation somewhere later in this section
(BTW, does checking for the optional relocations matter?  I guess it saves a couple extra strcmp's)

Adding "__might_sleep" to the relocation check works out for **generic_make_request_checks()** but I'm wondering if this algorithm is a bit optimistic.  I'm trying to think of an example patch that could inadvertently pass the test, but it's getting late on Friday afternoon and I'm stumped :)

I'd be curious to see other patterns that you may had noticed.